doc/reference.md

```diff
@@ -0,0 +1,509 @@
+
+Strings can be constructed as literals, slices of existing strings, concatenations of existing strings or converted from array of Unicode codepoint numbers.
+
+### Number
+
+Jsonnet numbers are 64-bit floating point numbers as defined in IEEE754 excluding `nan` and `inf` values. Operation resulting in infinity or not a number is considered an error. 
+
+Integers can be precisely represented as a Jsonnet number in the range [-2^53,2^53]. This is [a direct consequence of IEEE754 spec](https://en.wikipedia.org/wiki/Double-precision_floating-point_format).
+
+### Function
+
+Functions in Jsonnet are functions in the mathematical sense. Each function has parameters and a body expression. The result of calling a function is equivalent to the result of evaluating its body with arguments introduced to the environment.
+
+The basic way to define a Jsonnet function, is using a *function literal*:
+```
+local func = function(x) x * 2;
+func(21)
+```
+
+Jsonnet has syntax sugar for defining functions, which allows shorter, equivalent version:
+```
+local func(x) = x * 2;
+func(21)
+```
+
+The arguments are **not** evaluated when a function is called. They are passed lazily and evaluated only when actually used. This allows expressing things, which in other languages require built-in functionality or macros, such as the short-circuiting boolean operations:
+```
+local and3(a, b, c) = a && b && c;
+and3(true, false, error "this one is never evaluated")
+```
+
+It is worth noting at this point that functions in Jsonnet are [referentially transparent](Referential_transparency), meaning that any function call can be replaced with its definition, without changing the meaning of the program. Therefore, in some sense, functions in Jsonnet are "hygienic macros". For example consider the following snippet:
+
+```
+local pow2(n) = if n = 0 then 1 else 2 * pow2(n);
+pow2(17)
+```
+
+The call to function pow2 can be replaced with its definition as follows:
+
+```
+local pow2(n) = if n = 0 then 1 else 2 * pow2(n);
+local n = 17;
+if n = 0 then 1 else 2 * pow2(n)
+```
+
+#### Function Parameters
+
+The parameters can be either required or optional. In the definition, required and optional parameters can be mixed in any order. Optional parameters require specifying the default argument.
+
+When calling a function, each argument can be passed either as named or as positional, but all positional arguments need to go before the named arguments.
+
+For example the following program is valid: 
+
+```
+local foo(x, y=1) = x + y;
+[
+  foo(1),
+  foo(1, 1),
+  foo(x=1, y=1),
+  foo(y=1, x=1),
+  foo(x=1),
+]
+```
+
+It is recommended to always pass optional arguments as named (for readability). We also recommend to pass arguments as positional when all parameters are required, because the function author probably did not consider the names of the function names as part of the stable interface.
```

Comment:
shouldn't this read: ... because the function author probably did not consider the names of `the arguments` of the function as part ...

Reply:
Done.

Reply:
Historically using the named arguments where all parameters were required caused trouble – e.g. sometimes the names of the parameters often didn't actually match the documentation, even in the stdlib. And there's still https://github.com/google/jsonnet/issues/385.